Proposal: Make OCSP Optional, Require CRLs, and Incentivize Automation / Short-Lived Certificates (#414)
cabforum/servercert: docs/BR.md

---

> title: Baseline Requirements for the Issuance and Management of Publicly-Trusted Certificates

Reviewer: ryancdickson:make-ocsp-optional

**Root Certificate**: The self-signed Certificate issued by the Root CA to identify itself and to facilitate verification of Certificates issued to its Subordinate CAs.

> **Short-lived Subscriber Certificate**: A server Certificate with a Validity Period that is less than ten (10) days (864,000 seconds). 

Author: On #402, @BenWilson-Mozilla commented...

```I would add "that is" so that it reads, "with a Validity Period that is less than ..."```

This change has been incorporated.

> **Short-lived Subscriber Certificate**: A server Certificate with a Validity Period that is less than ten (10) days (864,000 seconds). 

Author: In #402, @BenWilson-Mozilla commented:

```Can the Validity Period be defined as a shorter period, e.g. five days, 432,000 seconds?```

Open for discussion. 

10 days was in attempt to:
- align with existing timelines (e.g.,  CRL "nextUpdate") and implementations observed in some browsers  
- offer an attractive incentive for CAs to adopt automation. Reducing the period too far may result in limited adoption due to the impact of unforeseen outages or operational issues. IMO, a validity of 10 days is more "forgiving" of an unexpected 2-3 day service outage than a validity of 5 days).

What are your thoughts on a phased approach (sample below):
- Until 2026: 10 days
- Effective 1/1/26: 5 days

Author: @BenWilson-Mozilla - added a phased reduction in 6.3.2.

```Subscriber Certificates issued on or after 1 September 2020 SHOULD NOT have a Validity Period greater than 397 days and MUST NOT have a Validity Period greater than 398 days. To qualify as a Short-lived Subscriber Certificate, Subscriber Certificates issued after 15 January 2024, but prior to 15 January 2026, MUST NOT have a Valdity Period greater than 10 days. Short-lived Subscriber Certificates issued on or after 15 January 2026 MUST NOT have a Validity Period greater than 7 days.```

Author: Closing stale comments.

#### 4.9.1.1 Reasons for Revoking a Subscriber Certificate

The CA MAY support revocation of Short-lived Subscriber Certificates.

> With the exception of Short-lived Subscriber Certificates, the CA SHALL revoke a Certificate within 24 hours if one or more of the following occurs:

Author: Closing stale comments.

1. The Subscriber requests in writing that the CA revoke the Certificate;
2. The Subscriber notifies the CA that the original certificate request was not authorized and does not retroactively grant authorization;
3. The CA obtains evidence that the Subscriber's Private Key corresponding to the Public Key in the Certificate suffered a Key Compromise;
4. The CA is made aware of a demonstrated or proven method that can easily compute the Subscriber's Private Key based on the Public Key in the Certificate (such as a Debian weak key, see <https://wiki.debian.org/SSLkeys>);
5. The CA obtains evidence that the validation of domain authorization or control for any Fully-Qualified Domain Name or IP address in the Certificate should not be relied upon.

> With the exception of Short-lived Subscriber Certificates, the CA SHOULD revoke a certificate within 24 hours and MUST revoke a Certificate within 5 days if one or more of the following occurs:

Author: On #402, @BenWilson-Mozilla commented:

```Same comment about 10-day validity period for short-lived certificates vs. 24-hour or 5-day revocation (assuming that revocation can cause a TLS connection to hard-fail).```

In response, please see my comment above. [Open for continued discussion!]

Author: Closing stale comments.

**Note**: Following certificate issuance, a certificate may be revoked for reasons stated in [Section 4.9](#49-certificate-revocation-and-suspension). Therefore, relying parties should check the revocation status of all certificates that contain a CDP or OCSP pointer.

### 4.9.7 CRL issuance frequency

CAs MUST generate and publish either:
- a full and complete; OR 
- partitioned (i.e., "sharded") CRLs, that when aggregated, represent the equivalent of the full and complete CRL.

CRLs must be available via a publicly-accessible HTTP URL. 

CAs issuing Subscriber Certificates (i.e., Certificates for servers) SHALL:

> - update and reissue CRLs at least once daily; and

Reviewer 2: If the intent is indeed to move from "at least once every seven days", to "at least daily", should we rephrase this to "at least once every 24 hours"?

> - the value of the `nextUpdate` field MUST NOT be more than ten (10) days beyond the value of the `thisUpdate` field.

Author: On #402, @BenWilson-Mozilla commented:

```Can this 10-day window be shortened?```

In part, the validity of the as-proposed "short-lived certificate" was intended to align with this timeline (i.e., 10 days) as it represented how long a client might cache and rely on a CRL.

Given CAs would be obligated to generate and publish new CRLs every 24 hours ("at least once daily"), applications like CRLSet/CRLite/OneCRL would be capable of obtaining fresh revocation status information every 24 hours, regardless of the nextUpdate.

Given the clarification above, do you still have concern about the 10-day timeline?

Author: Closing stale comments.

2. Be signed by an OCSP Responder whose Certificate is signed by the CA that issued the Certificate whose revocation status is being checked.

In the latter case, the OCSP signing Certificate MUST contain an extension of type `id-pkix-ocsp-nocheck`, as defined by RFC6960.

### 4.9.10 On-line revocation checking requirements

The CA MAY support on-line revocation status checking using OCSP. 

If the CA supports OCSP, the following requirements SHALL apply:

OCSP responders operated by the CA SHALL support the HTTP GET method, as described in RFC 6960 and/or RFC 5019.

The validity interval of an OCSP response is the difference in time between the `thisUpdate` and `nextUpdate` field, inclusive. For purposes of computing differences, a difference of 3,600 seconds shall be equal to one hour, and a difference of 86,400 seconds shall be equal to one day, ignoring leap-seconds.

For the status of Subscriber Certificates:

- Prior to 2020-09-30:
     1. The CA SHALL update information provided via an Online Certificate Status Protocol at least every four days. 
     2. OCSP responses from this service MUST have a maximum expiration time of ten days.

> - Effective 2020-09-30:

Author: On #402, @BenWilson-Mozilla commented:

```Should this be shortened, too?```

Open for community discussion. This proposal does not change the existing OCSP requirements, it merely makes OCSP optional.

Author: Closing stale comments.

**Note**: Following certificate issuance, a certificate may be revoked for reasons stated in [Section 4.9](#49-certificate-revocation-and-suspension). Therefore, relying parties should check the revocation status of all certificates that contain a CDP or OCSP pointer.

### 4.9.7 CRL issuance frequency

CAs MUST generate and publish either:
- a full and complete; OR
- partitioned (i.e., "sharded") CRLs, that when aggregated, represent the equivalent of the full and complete CRL.

CRLs must be available via a publicly-accessible HTTP URL.

> CAs issuing Subscriber Certificates (i.e., Certificates for servers) SHALL:

Author: Overcome by events, modified presentation format in https://github.com/ryancdickson/staging/blob/make-ocsp-optional-updates/docs/BR.md#497-crl-issuance-frequency. Will close this comment. Future comments should be made on https://github.com/ryancdickson/staging/pull/3 which will serve as the basis for the next round of public discussion.

- update and reissue CRLs at least once every 24 hours; and
- the value of the `nextUpdate` field MUST NOT be more than ten (10) days beyond the value of the `thisUpdate` field.

> CAs issuing CA Certificates SHALL:

Author: Same as response to prior comment. (OBE).

**Note**: Following certificate issuance, a certificate may be revoked for reasons stated in [Section 4.9](#49-certificate-revocation-and-suspension). Therefore, relying parties should check the revocation status of all certificates that contain a CDP or OCSP pointer.

### 4.9.7 CRL issuance frequency

CRLs must be available via a publicly-accessible HTTP URL (i.e., "published").

Within twenty-four (24) hours of issuing its first Certificate, the CA MUST generate and publish either:
- a full and complete CRL; OR
- partitioned (i.e., "sharded") CRLs that, when aggregated, represent the equivalent of a full and complete CRL.

CAs issuing Subscriber Certificates:  
1. MUST update and publish a new CRL within at least: 
     - seven (7) days if all Certificates include an Authority Information Access extension with an id-ad-ocsp accessMethod (“AIA OCSP pointer”); or

>      - four (4) days in all other cases; 

Reviewer 3: If we don't use "**at least** every X days" then if a CA decides to publish a new CRL sooner it would violate the requirement.

>      - four (4) days in all other cases; 

Reviewer 4: Ah good catch, then I'm in favor of replacing "within at least" with "at least every".

>      - four (4) days in all other cases; 

Author: @aarongable and @dzacharo - any issues with [this](https://github.com/ryancdickson/staging/commit/eb974081f456e73e68fd6df5382efabf4bc898f4) proposed update?

2. MUST update and publish a new CRL within twenty-four (24) hours after recording a Certificate as revoked; and

> 3. MUST include a `nextUpdate` field value that is no more than ten (10) days beyond the value of the `thisUpdate` field.

Author: I appreciate that perspective. 

I originally kept the language in 4.9.7 for consistency with the existing location in the BRs. However, now that this requirement is addressed in the CRL profile, the redundancy may create more headache than it's worth (i.e., if the nextUpdate period were to change in the future, we'd need to address it in two sections).

Addressed in: https://github.com/ryancdickson/staging/commit/96153723ca783676a1b1f0316bc8610435b6c79b

### 4.10.2 Service availability

> The CA SHALL operate and maintain its CRL and optional OCSP capability with resources sufficient to provide a response time of ten seconds or less under normal operating conditions.

Author: In #402, @BenWilson-Mozilla commented:

I think there has been discussion about eliminating or re-phrasing this requirement.

[will leave open for community discussion]

Author: Closing stale comments.

### 6.3.2 Certificate operational periods and key pair usage periods

> Subscriber Certificates issued on or after 1 September 2020 SHOULD NOT have a Validity Period greater than 397 days and MUST NOT have a Validity Period greater than 398 days. To qualify as a Short-lived Subscriber Certificate, Subscriber Certificates issued after 15 March 2024, but prior to 15 March 2026, MUST NOT have a Valdity Period greater than 10 days. Short-lived Subscriber Certificates issued on or after 15 March 2026 MUST NOT have a Validity Period greater than 7 days.

Author: Addressed in https://github.com/ryancdickson/staging/commit/df5e689b2b098b21aa997299b1c8dd1b927a74f4.

| __Access Method__ | __OID__            | __Access Location__         | __Presence__ | __Maximum__ | __Description__ |
| --                | --                 | ----                        | -            | -          | ---             |

> | `id-ad-ocsp`      | 1.3.6.1.5.5.7.48.1 | `uniformResourceIdentifier` | MAY       | \*         | A HTTP URL of the Issuing CA's OCSP responder. |

Reviewer 4: nit: column alignment
```suggestion
| `id-ad-ocsp`      | 1.3.6.1.5.5.7.48.1 | `uniformResourceIdentifier` | MAY          | \*         | A HTTP URL of the Issuing CA's OCSP responder. |
```

##### 7.1.2.11.2 CRL Distribution Points

The CRL Distribution Points extension MUST be present in:
- Subordinate CA Certificates; and 

> - Subscriber Certificates with validity greater than or equal to ten days.

Author: In #402, @BenWilson-Mozilla commented:

```Change to 5 days, 432,000 seconds?```

Open for discussion (described in open comment above). This value will be set based on the validity of a short-lived certificate.

Author: Closing stale comments.

##### 7.1.2.11.2 CRL Distribution Points

The CRL Distribution Points extension MUST be present in:
- Subordinate CA Certificates; and
- Subscriber Certificates that 1) do not qualify as "Short-lived Subscriber Certificates" and 2) do not include an Authority Information Access extension with an id-ad-ocsp accessMethod.

The CRL Distribution Points extension SHOULD NOT be present in:
- Root CA Certificates.

The CRL Distribution Points extension is OPTIONAL in:
- Short-lived Subscriber Certificates.

The CRL Distribution Points extension MUST NOT be present in:
- OCSP Responder Certificates.

> When present, the CRL Distribution Points extension MUST contain at least one `DistributionPoint`; containing more than one is NOT RECOMMENDED. All `DistributionPoint` items must be formatted as follows:

Reviewer 4: ```suggestion
When present, the CRL Distribution Points extension MUST contain at least one `DistributionPoint`; containing more than one is NOT RECOMMENDED. All `DistributionPoint` items MUST be formatted as follows:
```
maybe? I'm not sure why this "must" wasn't capitalized previously.

Author: It was a typo on my part dating back to the [early versions](https://github.com/ryancdickson/staging/blob/36bd26adf1346fcb9ff23e43468501778a7d3d7c/docs/BR.md#:~:text=When%20present%2C%20the%20CRL%20Distribution%20Points%20extension%20MUST%20contain%20at%20least%20one%20DistributionPoint%3B%20containing%20more%20than%20one%20is%20NOT%20RECOMMENDED.%20All%20DistributionPoint%20items%20must%20be%20formatted%20as%20follows%3A) of the proposal.

##### 7.1.2.11.2 CRL Distribution Points

The CRL Distribution Points extension MUST be present in:
- Subordinate CA Certificates; and
- Subscriber Certificates that 1) do not qualify as "Short-lived Subscriber Certificates" and 2) do not include an Authority Information Access extension with an id-ad-ocsp accessMethod.

The CRL Distribution Points extension SHOULD NOT be present in:
- Root CA Certificates; and

> - Short-lived Subscriber Certificates.

Reviewer 5: I think "SHOULD NOT be present in...Short-lived Subscriber Certificates" is a step too far.

CAs that issue <=398-day certificates may sometimes issue <10-day certificates, but not because they want to avoid having to provide revocation information.  For example: if a Subscriber suffers a key compromise 7 days before their certificate is due to expire, the CA might issue a 7-day replacement certificate (with a different public key) that will expire at the same time, in order to cover (but not exceed) the validity period that the Subscriber had already paid for.

```suggestion
The CRL Distribution Points extension is OPTIONAL in:
- Short-lived Subscriber Certificates.
```

Author: No concerns with this recommendation. 

Addressed in [this](https://github.com/ryancdickson/staging/pull/5/files) PR.

## 7.2 CRL profile

Prior to 2024‐03‐15, the CA SHALL issue CRLs in accordance with the profile specified in these Requirements or the profile specified in Version 1.8.7 of the Baseline Requirements for the Issuance and Management of Publicly‐Trusted Certificates. Effective 2024‐03‐15, the CA SHALL issue CRLs in accordance with the profile specified in these Requirements.

### 7.2.1 Version number(s)

Certificate Revocation Lists MUST be of type X.509 v2.

### 7.2.2 CRL and CRL entry extensions

If the CA asserts compliance with these Baseline Requirements, all CRLs that it issues MUST comply with the following CRL profile, which incorporates, and is derived from [RFC 5280](https://tools.ietf.org/html/rfc5280). Except as explicitly noted, all normative requirements imposed by RFC 5280 shall apply, in addition to the normative requirements imposed by this document. CAs SHOULD examine [RFC 5280, Appendix B](https://tools.ietf.org/html/rfc5280#appendix-B) for further issues to be aware of.

> A full and complete CRL is a list of all revoked certificates issued by the CA for any and all reasons.
> 
> A partitioned CRL (sometimes referred to as a "sharded CRL") is a list of revoked certificates issued by the CA for any and all reasons constrained to a specific scope (e.g., temporal sharding).
> 
> Minimally, CAs MUST issue either a "full and complete" CRL - or "partitioned" CRLs. Aside from the presence of the `IssuingDistributionPoint` extension (OID 2.5.29.28) in partitioned CRLs, both CRL formats are syntactically the same from the perspective of this profile.
> 
> If using only partitioned CRLs, the full set of partitioned CRLs MUST cover the complete set of public-key certificates issued by the CA. Thus, the complete set of partitioned CRLs MUST be equivalent to a full CRL for the same set of public-key certificates, if the CA was not using partitioned CRLs. 

Reviewer 4: This is wildly editorial, so feel free to ignore, but I would combine/rephrase/reorder these paragraphs. I works better for my brain this way, but that's definitely not a claim that my way works better for everyone or even a majority of people:

```suggestion
A full and complete CRL is CRL whose scope includes all certificates issued by the CA.

A partitioned CRL (sometimes referred to as a "sharded CRL") is a CRL with a a constrained scope, such as all certificates issued by the CA during a certain period of time ("temporal sharding"). Aside from the presence of the `IssuingDistributionPoint` extension (OID 2.5.29.28) in partitioned CRLs, both CRL formats are syntactically the same from the perspective of this profile.

Minimally, CAs MUST issue either a "full and complete" CRL or a set of "partitioned" CRLs which cover the complete set of certificates issued by the CA. In other words, if issuing only partitioned CRLs, the combined scope of those CRLs must be equivalent to that of a full and complete CRL. 
```

Author: Accepted suggestion with minor tweaks.  https://github.com/ryancdickson/staging/commit/12c2db11346481fd8efd150144692950827f968f

CAs MUST NOT issue indirect CRLs (i.e., the issuer of the CRL is not the issuer of all Certificates that are included in the scope of the CRL).

Table: CRL Fields

| __Field__                  | __Presence__    | __Description__ |
| ---                        | ------          | ------          |
| `tbsCertList`              |                 |                 |
|     `version`              | MUST     | MUST be v2(1) |
|     `signature`            | MUST     | See [Section 7.1.3.2](#7132-signature-algorithmidentifier) |
|     `issuer`               | MUST     | MUST be byte-for-byte identical to the `subject` field of the Issuing CA. |
|     `thisUpdate`           | MUST     | UTCTime (YYMMDDHHMMSSZ) MUST be used for dates up to and including 2049. GeneralizedTime (YYYYMMDDHHMMSSZ) MUST be used for dates after 2049.|
|     `nextUpdate`           | MUST     | UTCTime (YYMMDDHHMMSSZ) MUST be used for dates up to and including 2049. GeneralizedTime (YYYYMMDDHHMMSSZ) MUST be used for dates after 2049.|

> |     `revokedCertificates`  | *        | MUST only be present if the CA has issued a certificate that is both revoked and unexpired. An entry MUST NOT be removed from the CRL until it appears on one regularly scheduled CRL issued beyond the revoked certificate's validity period. See the "revokedCertificates Component" table for additional requirements. |

Reviewer 4: This description is too strict. CRL entries are required to remain on a CRL for one CRL issuance cycle _after_ they expire (so that a cert which is revoked seconds before it expires appears on at least one CRL anyway). This requirement would force a CA to violate _that_ requirement if the CA has only expired-and-revoked certificates left, which is a common scenario for the very last CRL published at the end of a CA's issuing lifetime.

Author: Tweaked in https://github.com/ryancdickson/staging/commit/5cc135f2eec7241dd3868369e50b200b641cf813.

"MUST be present if the CA has issued a certificate that has been revoked and the corresponding entry has yet to appear on at least one regularly scheduled CRL beyond the revoked certificate's validity period. The CA SHOULD remove an entry for a corresponding certificate after it has appeared on at least one regularly scheduled CRL beyond the revoked certificate's validity period."

Reviewer 6: Roots that are under both TLS BR and CS BR audit scope will need to violate the SHOULD for revoked code signing ICAs (CSBRs require CRL revocation services for 10 years after expiry), but I think that following the CSBRs is compelling justification for violating this SHOULD-level requirement.

|     `extensions`           | MUST     | See the "CRL Extensions" table for additional requirements. |
| `signatureAlgorithm`       | MUST     | Encoded value MUST be byte-for-byte identical to the `tbsCertList.signature`. |
| `signature`                | MUST     | - |
| Any other value            | NOT RECOMMENDED | - |

Table: CRL Extensions

| __Extension__                     | __Presence__    | __Critical__          | __Description__ |
| ----                              | -               | -                     | ----- |
| `authorityKeyIdentifier`          | MUST            | N                     | The `keyIdentifier` field MUST be included and its value MUST be byte-for-byte identical to the Subject Key Identifier of the Issuing CA's Certificate. <br><br>The `authorityCertIssuer` and `authorityCertSerialNumber` fields MUST NOT be populated.      |

> | `CRLNumber`                       | MUST            | N                     | MUST contain an INTEGER greater than or equal to zero (0) and less than 2¹⁵⁹, and convey a monotonically increasing sequence.       |

Reviewer 4: "Monotonic" just means non-decreasing; CRL Numbers are required to be _strictly_ increasing (they cannot stay the same between CRL generations).
```suggestion
| `CRLNumber`                       | MUST            | N                     | MUST contain an INTEGER greater than or equal to zero (0) and less than 2¹⁵⁹, and convey a strictly increasing sequence.       |
```

Author: Addressed in https://github.com/ryancdickson/staging/commit/f65d2470e4777136d567c5a4ff5efa42c216b206.

A full and complete CRL is a CRL whose scope includes all certificates issued by the CA.

A partitioned CRL (sometimes referred to as a "sharded CRL") is a CRL with a constrained scope, such as all certificates issued by the CA during a certain period of time ("temporal sharding"). Aside from the presence of the `IssuingDistributionPoint` extension (OID 2.5.29.28) in partitioned CRLs, both CRL formats are syntactically the same from the perspective of this profile.

Minimally, CAs MUST issue either a "full and complete" CRL or a set of "partitioned" CRLs which cover the complete set of certificates issued by the CA. In other words, if issuing only partitioned CRLs, the combined scope of those CRLs must be equivalent to that of a full and complete CRL. 

CAs MUST NOT issue indirect CRLs (i.e., the issuer of the CRL is not the issuer of all Certificates that are included in the scope of the CRL).

Table: CRL Fields

> | __Field__                  | __Presence__    | __Description__ |
> | ---                        | ------          | ------          |
> | `tbsCertList`              |                 |                 |
> |     `version`              | MUST     | MUST be v2(1), see [Section 7.2.1](#721-version-numbers) |
> |     `signature`            | MUST     | See [Section 7.1.3.2](#7132-signature-algorithmidentifier) |
> |     `issuer`               | MUST     | MUST be byte-for-byte identical to the `subject` field of the Issuing CA. |
> |     `thisUpdate`           | MUST     | Indicates the issue date of the CRL. |
> |     `nextUpdate`           | MUST     | Indicates the date by which the next CRL will be issued. For CRLs covering Subscriber Certificates, at most 10 days after the `thisUpdate`. For other CRLs, at most 12 months after the `thisUpdate`. |
> |     `revokedCertificates`  | *        | MUST be present if the CA has issued a certificate that has been revoked and the corresponding entry has yet to appear on at least one regularly scheduled CRL beyond the revoked certificate's validity period. The CA SHOULD remove an entry for a corresponding certificate after it has appeared on at least one regularly scheduled CRL beyond the revoked certificate's validity period. See the "revokedCertificates Component" table for additional requirements.  |
> |     `extensions`           | MUST     | See the "CRL Extensions" table for additional requirements. |
> | `signatureAlgorithm`       | MUST     | Encoded value MUST be byte-for-byte identical to the `tbsCertList.signature`. |
> | `signature`                | MUST     | - |
> | Any other value            | NOT RECOMMENDED | - |

Reviewer 4: nit: column alignment
```suggestion
| __Field__                  | __Presence__    | __Description__ |
| ---                        | ------          | ------          |
| `tbsCertList`              |                 |                 |
|     `version`              | MUST            | MUST be v2(1), see [Section 7.2.1](#721-version-numbers) |
|     `signature`            | MUST            | See [Section 7.1.3.2](#7132-signature-algorithmidentifier) |
|     `issuer`               | MUST            | MUST be byte-for-byte identical to the `subject` field of the Issuing CA. |
|     `thisUpdate`           | MUST            | Indicates the issue date of the CRL. |
|     `nextUpdate`           | MUST            | Indicates the date by which the next CRL will be issued. For CRLs covering Subscriber Certificates, at most 10 days after the `thisUpdate`. For other CRLs, at most 12 months after the `thisUpdate`. |
|     `revokedCertificates`  | *               | MUST be present if the CA has issued a certificate that has been revoked and the corresponding entry has yet to appear on at least one regularly scheduled CRL beyond the revoked certificate's validity period. The CA SHOULD remove an entry for a corresponding certificate after it has appeared on at least one regularly scheduled CRL beyond the revoked certificate's validity period. See the "revokedCertificates Component" table for additional requirements.  |
|     `extensions`           | MUST            | See the "CRL Extensions" table for additional requirements. |
| `signatureAlgorithm`       | MUST            | Encoded value MUST be byte-for-byte identical to the `tbsCertList.signature`. |
| `signature`                | MUST            | - |
| Any other value            | NOT RECOMMENDED | - |
```

Author: Addressed in https://github.com/ryancdickson/staging/commit/bbf5155dd87de476eb360a79feb8e49fdd0eefe6.

### 7.2.1 Version number(s)

Certificate Revocation Lists MUST be of type X.509 v2.

### 7.2.2 CRL and CRL entry extensions

Table: CRL Extensions

> | __Extension__                     | __Presence__    | __Critical__          | __Description__ |
> | ----                              | -               | -                     | ----- |
> | `authorityKeyIdentifier`          | MUST            | N                     | See [Section 7.1.2.11.1](#712111-authority-key-identifier) |
> | `CRLNumber`                       | MUST            | N                     | MUST contain an INTEGER greater than or equal to zero (0) and less than 2¹⁵⁹, and convey a strictly increasing sequence.        |
> | `IssuingDistributionPoint`        | *           | Y                     | See [Section 7.2.2.1 CRL Issuing Distribution Point](#7221-crl-issuing-distribution-point) |
> | Any other extension               | NOT RECOMMENDED        | -                     |       |

Reviewer 4: nit: column alignment
```suggestion
| __Extension__                     | __Presence__    | __Critical__          | __Description__ |
| ----                              | -               | -                     | ----- |
| `authorityKeyIdentifier`   | MUST            | N                     | See [Section 7.1.2.11.1](#712111-authority-key-identifier) |
| `CRLNumber`                | MUST            | N                     | MUST contain an INTEGER greater than or equal to zero (0) and less than 2¹⁵⁹, and convey a strictly increasing sequence.        |
| `IssuingDistributionPoint` | *               | Y                     | See [Section 7.2.2.1 CRL Issuing Distribution Point](#7221-crl-issuing-distribution-point) |
| Any other extension        | NOT RECOMMENDED | -                     |       |
```

Author: Addressed in https://github.com/ryancdickson/staging/commit/b2df5dc24cdb9333ffa166edfe10c09445b3eeff.

Table: revokedCertificates Component

> | __Component__                     | __Presence__    | __Description__ |
> | ----                              | -               | ----- |
> | `serialNumber`                    | MUST            | MUST be byte-for-byte identical to the serialNumber contained in the revoked certificate. |
> | `revocationDate`                  | MUST            | Normally, the date and time revocation occurred. See the footnote following this table for circumstances where backdating is permitted.  | 
> | `crlEntryExtensions`              | *               | See the "crlEntryExtensions Component" table for additional requirements. |

Reviewer 4: nit: column width
```suggestion
| __Component__        | __Presence__ | __Description__ |
| ----                              | -               | ----- |
| `serialNumber`       | MUST         | MUST be byte-for-byte identical to the serialNumber contained in the revoked certificate. |
| `revocationDate`     | MUST         | Normally, the date and time revocation occurred. See the footnote following this table for circumstances where backdating is permitted.  | 
| `crlEntryExtensions` | *            | See the "crlEntryExtensions Component" table for additional requirements. |
```

Author: Addressed in https://github.com/ryancdickson/staging/commit/bbf5155dd87de476eb360a79feb8e49fdd0eefe6.

Table: crlEntryExtensions Component 

| __CRL Entry Extension__   | __Presence__  | __Description__ |
| ---                       | -             | ------          |
| `reasonCode`              | *             | When present (OID 2.5.29.21), MUST NOT be marked critical and MUST indicate the most appropriate reason for revocation of the Certificate. <br><br> MUST be present unless the CRL entry is for a Certificate not technically capable of causing issuance and either 1) the CRL entry is for a Subscriber Certificate subject to these Requirements revoked prior to July 15, 2023 or 2) the reason for revocation (i.e., reasonCode) is unspecified (0). <br><br>See the "CRLReasons" table for additional requirements. |
| Any other value | NOT RECOMMENDED | |

Table: CRLReasons

| __RFC 5280 reasonCode__   | __RFC 5280 reasonCode value__ | __Description__ |
| ---                       | -    | ------          |

> | unspecified               | 0    |  MUST be omitted if the CRL entry is for a Certificate not technically capable of causing issuance unless the CRL entry is for a Subscriber Certificate subject to these Requirements revoked prior to July 15, 2023. 

Reviewer 7: Nit, but I had to read this language a few times to get comfortable with it. I think a statement like 'the unspecified reason is represented by the omission of a reasonCode.' would add clarity.

Author: Addressed in https://github.com/ryancdickson/staging/commit/530bc887dfd9553d8a77e1de5c5f20e1ed9d67b1

The "Description" column now reads: "_Represented by the omission of a reasonCode. MUST be omitted if the CRL entry is for a Certificate not technically capable of causing issuance unless the CRL entry is for a Subscriber Certificate subject to these Requirements revoked prior to July 15, 2023._"

@wthayer, does this work for you?

Reviewer 7: LGTM

The Subscriber Agreement, or an online resource referenced therein, MUST inform Subscribers about the revocation reason options listed above and provide explanation about when to choose each option. Tools that the CA provides to the Subscriber MUST allow for these options to be easily specified when the Subscriber requests revocation of their Certificate, with the default value being that no revocation reason is provided (i.e. the default corresponds to the CRLReason “unspecified (0)” which results in no reasonCode extension being provided in the CRL). 

The privilegeWithdrawn reasonCode SHOULD NOT be made available to the Subscriber as a revocation reason option, because the use of this reasonCode is determined by the CA and not the Subscriber.

When a CA obtains verifiable evidence of Key Compromise for a Certificate whose CRL entry does not contain a reasonCode extension or has a reasonCode extension with a non-keyCompromise reason, the CA SHOULD update the CRL entry to enter keyCompromise as the CRLReason in the reasonCode extension. 

> #### 7.2.2.1 CRL Issuing Distribution Point

Reviewer 4: nit: empty line after header
```suggestion
#### 7.2.2.1 CRL Issuing Distribution Point

```

Partitioned CRLs MUST include at least one of the names from the corresponding distributionPoint field of the cRLDistributionPoints extension of every certificate that is within the scope of this CRL. The encoded value MUST be byte-for-byte identical to the encoding used in the distributionPoint field of the certificate. 

> 

Reviewer 7: What about CAs that shard CRLs but do not include the cRLDP in the certs? I don't think the intent is to prohibit that but this language seems to have that effect? Can we add something like 'or at least one of the URLs published in CCADB'? Also, in either case is there a chicken/egg problem here if CRLs are issued before certs or before CCADB is updated?

Reviewer 7: My fundamental concern is that someone could come along in the future and interpret this section as an implicit requirement to include the cRLDP extension in certs when a CA is issuing sharded CRLs. Perhaps an easy clarification that aligns with your interpretation of the language is to prepend "If certificates contain a distributionPoints field and that name points to a partitioned CRL,..."

Author: Fair point. Re-phrased, with that concern in mind:

"_If a Certificate contains a distributionPoint field that does not point to the URI of a full and complete CRL, the corresponding CRL MUST contain the issuingDistributionPoint extension (OID 2.5.29.28). The encoded value of the issuingDistributionPoint extension MUST be byte-for-byte identical to the encoding used in the distributionPoint field of the Certificate._"

Any concerns with this approach?

Reviewer 7: LGTM

Author: Thanks @wthayer!

CC @aarongable and @CBonnell - as I believe during past discussions you both had opinions on the language re: IDP. Any concerns with the above change? Candidly, I'd like this next round of discussion to be the last, so the more opinions, the better.

Reviewer 6: I understand the intent behind the change, but the concern I have with the proposed language is that other GeneralNames cannot be included in the IDP extension (as the value of the extension itself must be binary equal to the single distributionPoint).

I suggest:

Partitioned CRLs MUST contain an Issuing Distribution Point extension. The `distributionPoint` field of the Issuing Distribution Point extension MUST be present. Additionally, the `fullName` field of the DistributionPointName value MUST be present, and its value MUST conform to the following requirements:

1. If a Certificate within the scope of the CRL contains a CRL Distribution Points extension, then at least one of the `uniformResourceIdentifier`s in the CRL Distribution Points's `fullName` field MUST be included in the `fullName` field of the CRL's Issuing Distribution Point extension. The encoding of the `uniformResourceIdentifier` value in the Issuing Distribution Point extension SHALL be identical to the encoding used in the Certificate's CRL Distribution Points extension.
2. Other GeneralNames of type `uniformResourceIdentifier` MAY be included.
3. Non-`uniformResourceIdentifier` GeneralName types MUST NOT be included.

Author: Thanks, @CBonnell. I think that works and still addresses @wthayer's earlier concern.

Aside from one small change (using "byte-for-byte" in lieu of "octet-for-octet" --- to avoid introducing a new term), I believe I [incorporated](https://github.com/ryancdickson/staging/commit/15eab80d50a7b6b0e5dce7d413c7692610d4ac93) your feedback "as-is." 

I made some other small tweaks to improve consistency in how extensions introduced in this update are represented.

Minimally, CAs MUST issue either a "full and complete" CRL or a set of "partitioned" CRLs which cover the complete set of certificates issued by the CA. In other words, if issuing only partitioned CRLs, the combined scope of those CRLs must be equivalent to that of a full and complete CRL. 

CAs MUST NOT issue indirect CRLs (i.e., the issuer of the CRL is not the issuer of all Certificates that are included in the scope of the CRL).

Table: CRL Fields

| __Field__                  | __Presence__    | __Description__ |
| ---                        | ------          | ------          |
| `tbsCertList`              |                 |                 |
|     `version`              | MUST     | MUST be v2(1) |
|     `signature`            | MUST     | See [Section 7.1.3.2](#7132-signature-algorithmidentifier) |
|     `issuer`               | MUST     | MUST be byte-for-byte identical to the `subject` field of the Issuing CA. |
|     `thisUpdate`           | MUST     | UTCTime (YYMMDDHHMMSSZ) MUST be used for dates up to and including 2049. GeneralizedTime (YYYYMMDDHHMMSSZ) MUST be used for dates after 2049.|

> |     `nextUpdate`           | MUST     | UTCTime (YYMMDDHHMMSSZ) MUST be used for dates up to and including 2049. GeneralizedTime (YYYYMMDDHHMMSSZ) MUST be used for dates after 2049.|

Reviewer 4: This language works for me.

|     `revokedCertificates`  | *        | MUST be present if the CA has issued a certificate that has been revoked and the corresponding entry has yet to appear on at least one regularly scheduled CRL beyond the revoked certificate's validity period. The CA SHOULD remove an entry for a corresponding certificate after it has appeared on at least one regularly scheduled CRL beyond the revoked certificate's validity period. See the "revokedCertificates Component" table for additional requirements.  |
|     `extensions`           | MUST     | See the "CRL Extensions" table for additional requirements. |
| `signatureAlgorithm`       | MUST     | Encoded value MUST be byte-for-byte identical to the `tbsCertList.signature`. |
| `signature`                | MUST     | - |

> | Any other value            | NOT RECOMMENDED | - |

Author: Addressed in: https://github.com/ryancdickson/staging/commit/c2dadd3a97c3b68a9a7363ed78c6c279ba89ca75

Table: CRL Extensions

| __Extension__                     | __Presence__    | __Critical__          | __Description__ |
| ----                              | -               | -                     | ----- |

> | `authorityKeyIdentifier`          | MUST            | N                     | The `keyIdentifier` field MUST be included and its value MUST be byte-for-byte identical to the Subject Key Identifier of the Issuing CA's Certificate. <br><br>The `authorityCertIssuer` and `authorityCertSerialNumber` fields MUST NOT be populated.      |

Author: Addressed in: https://github.com/ryancdickson/staging/commit/7767d3da5eabcdfec13a6a7bd2975a7dd6bcf14e

| `CRLNumber`                       | MUST            | N                     | MUST contain an INTEGER greater than or equal to zero (0) and less than 2¹⁵⁹, and convey a strictly increasing sequence.        |

> | `IssuingDistributionPoint`        | *           | Y                     | Partitioned CRLs MUST include at least one of the names from the corresponding distributionPoint field of the cRLDistributionPoints extension of every certificate that is within the scope of this CRL. The encoded value MUST be byte-for-byte identical to the encoding used in the distributionPoint field of the certificate. <br><br>The `indirectCRL` and `onlyContainsAttributeCerts` fields MUST be set to `FALSE` (i.e., not asserted). <br><br>The CA MAY set either of the `onlyContainsUserCerts` and `onlyContainsCACerts` fields to `TRUE`, depending on the scope of the CRL. <br><br>The CA MUST NOT assert both of the `onlyContainsUserCerts` and `onlyContainsCACerts` fields. <br><br>The `onlySomeReasons` field SHOULD NOT be included; if included, then the CA MUST provide another CRL whose scope encompasses all revocations regardless of reason code. <br><br>This extension is NOT RECOMMENDED for full and complete CRLs. |

Author: I had considered that, too. Thanks for the nudge. 

Addressed in https://github.com/ryancdickson/staging/commit/8d0022567dad16910c7486721403beb40b97b448

| Any other extension               | NOT RECOMMENDED        | -                     |       |

Table: revokedCertificates Component

| __Component__                     | __Presence__    | __Description__ |
| ----                              | -               | ----- |
| `serialNumber`                    | MUST            | MUST be byte-for-byte identical to the serialNumber contained in the revoked certificate. |

> | `revocationDate`                  | MUST            | The date and time which revocation occurred. UTCTime (YYMMDDHHMMSSZ) MUST be used for dates up to and including 2049. GeneralizedTime (YYYYMMDDHHMMSSZ) MUST be used for dates after 2049.  | 

Author: Agree - addressed in https://github.com/ryancdickson/staging/commit/09a563c73dfb4c33e285754a441a112d5507848c

| `crlEntryExtensions`              | *               | See the "crlEntryExtensions Component" table for additional requirements. |

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Reviewer 5: Also, I'm curious: Do there really exist "TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised"?  Or has this language accidentally been imported from the Code Signing BRs, which permit the RFC5280-non-compliant backdating of the revocationDate field as a workaround for the fact that Microsoft's stack doesn't support the RFC5280 [Invalidity Date](https://www.rfc-editor.org/rfc/rfc5280.html#section-5.3.2) extension?

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Reviewer 5: ISTM that TLS implementations only care about the revocation status of a certificate and pay no heed to any timestamps that indicate how far in the past the revocation occurred, whereas Code Signing does care about the "revocation time" and "invalidity date".

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Author: Thanks for this comment, @robstradling.

I attempted to address the backdating concerns [here](https://github.com/ryancdickson/staging/pull/5/files).

Rather than accepting your suggestion as written (i.e., move the complete paragraph beginning with "When a CA obtains..." to the beginning of the note) - I split the paragraph such that discussion re: CRLReasons follows the CRLReason table. Otherwise, we're introducing language related to CRLReasons that are not yet enumerated. 

Any concerns with how I handled this?

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Author: Re: your other comments - I agree with you re: "TLS implementations only care about the revocation status of a certificate and pay no heed to any timestamps that indicate how far in the past the revocation occurred".

It's not clear to me the motivation of the existing language re: backdating. @BenWilson-Mozilla - as the author of SC-61, can you offer any background?

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Reviewer 5: @ryancdickson That change looks good to me.  Thanks.

(I still think that the BRs should not be permitting the backdating of revocationDate, but tackling that existing language might be considered scope creep for this particular ballot).

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Author: Thanks for confirming, Rob!

I agree with your comment re: "scope creep," but will take action to create a separate issue such that we can explore this further in a future cleanup ballot.

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Author: Given it seems unlikely we'll get through the Ballot in 3 rounds of discussion due to some recent comments, I might explore this in a Version 4 (hopefully the last). 

I'll follow-up with @BenWilson-Mozilla to understand why the backdating discussion was pulled into the TLS BRs via SC-61. 

My preference would be to stipulate the revocation date listed on the CRL must be the _actual_ date of revocation. Any opinion on that @robstradling?

While studying the CRLs disclosed to CCADB, we noted a small number of instances where:
- Certificates appeared to be revoked before they were issued  ([example](https://crt.sh/?q=4e48fab0934f50cd3a42d00653089f1a90d56f6157cad82d9afce2e6dc81be4b))
- Certificates appeared to be revoked at the same time as when they were issued ([example](https://crt.sh/?id=7085490113))

As an outside observer, it's impossible to know the _actual_ revocation date - however we can gain some context  by observing when an entry lands on the CRL (not easily monitored using existing public tools).

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Author: Update: Discussed a bit with @CBonnell and @wthayer during lunch at the F2F. 

In cases where a pre-certificate is signed and the corresponding certificate is _never_ issued, setting revocationDate to equal the notBefore date makes sense. 

I'd still be interested in learning whether there are other valid use cases for backdating revocationDate.

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Reviewer 8: My understanding about the language re: backdating is that it was solely related to situations where the revocation reason code was being changed to keyCompromise from some other reason - to say that was allowed. Does that clarification help us navigate a path forward on this issue?

> **Note:** Backdating the revocationDate field is an exception to best practice described in RFC 5280 (Section 5.3.2); however, these requirements specify the use of the revocationDate field to support TLS implementations that process the revocationDate field as the date when the Certificate is first considered to be compromised.

Reviewer 4: Confirming Ben's understanding: The paragraph this comment thread is on is copied from [here](https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/#6-revocation:~:text=Note%3A%20Backdating%20the%20revocationDate%20field%20is%20an%20exception%20to%20best%20practice%20described%20in%20RFC%205280%20(section%205.3.2)) in the MRSP. That version contains a hyperlink to [here](https://wiki.mozilla.org/CA/Revocation_Reasons#Updating_CRL_Entries) which provides some clarifying language.
